include/oneapi/dpl/pstl/hetero/numeric_impl_hetero.h

```diff
@@ -108,23 +118,244 @@ __pattern_transform_reduce(_ExecutionPolicy&& __exec, _ForwardIterator __first,
+    static void __group_scan(const _Group& __group, _Begin __begin, _End __end, _BinaryOperation __bin_op, unseq_backend::__init_value<_ValueType> __init)
+    {
+        if constexpr (_Inclusive)
+            __dpl_sycl::__joint_inclusive_scan(__group, __begin, __end, __begin, __bin_op, __init.__value);
+        else
+            __dpl_sycl::__joint_exclusive_scan(__group, __begin, __end, __begin, __init.__value, __bin_op);
+    }
+};
+
+
 //------------------------------------------------------------------------
 // transform_scan
 //------------------------------------------------------------------------
-
 template <typename T>
 struct ExecutionPolicyWrapper;
 
-template <typename _ExecutionPolicy, typename _Iterator1, typename _Iterator2, typename _UnaryOperation,
+template <typename _ExecutionPolicy, typename _InRng, typename _OutRng, typename _UnaryOperation,
           typename _InitType, typename _BinaryOperation, typename _Inclusive>
 void
-__pattern_transform_scan_base_impl(_ExecutionPolicy&& __exec, _Iterator1 __first, _Iterator1 __last,
-                                   _Iterator2 __result, _UnaryOperation __unary_op, _InitType __init,
-                                   _BinaryOperation __binary_op, _Inclusive)
+__pattern_transform_scan_single_group(_ExecutionPolicy&& __exec, _InRng&& __in_rng, _OutRng __out_rng, ::std::size_t __n,
+                              _UnaryOperation __unary_op, _InitType __init, _BinaryOperation __binary_op, _Inclusive)
 {
-    const auto __n = __last - __first;
-    assert(__n > 0);
+    using _CustomName = typename std::decay_t<_ExecutionPolicy>::kernel_name;
+    using _DynamicGroupScanKernel =
+                 oneapi::dpl::__par_backend_hetero::__internal::__kernel_name_generator<__par_backend_hetero::__scan_single_wg_dynamic_kernel, _CustomName, _BinaryOperation, _InRng, _OutRng>;
 
+    ::std::size_t __max_wg_size = oneapi::dpl::__internal::__max_work_group_size(__exec);
+
+#if _ONEDPL_COMPILE_KERNEL
+    auto __kernel = __par_backend_hetero::__internal::__kernel_compiler<_DynamicGroupScanKernel>::__compile(::std::forward<_ExecutionPolicy>(__exec));
+    __max_wg_size = ::std::min(__max_wg_size, oneapi::dpl::__internal::__kernel_work_group_size(
+                                                          ::std::forward<_ExecutionPolicy>(__exec), __kernel));
+#endif
+
+    // Specialization for devices that have a max work-group size of 1024
+    constexpr ::std::uint16_t __targeted_wg_size = 1024;
+
+    if (__max_wg_size >= __targeted_wg_size)
+    {
+        auto __single_group_scan_f = [&](auto __size_constant) {
+            constexpr ::std::uint16_t __size = decltype(__size_constant)::value;
+            constexpr ::std::uint16_t __wg_size = ::std::min(__size, __targeted_wg_size);
+            constexpr ::std::uint16_t __num_elems_per_item = __par_backend_hetero::__ceiling_div(__size, __wg_size);
+            const bool __is_full_group = __n == __wg_size;
```

Comment:
Why do we rely on so rare case in "real life"? It looks just for good benchmark results reporting.....
When residual is empty, I believe the compiler gets rid of "residual" code.... or it doesn't so??

Reply:
Whether or not the residual is empty is a runtime condition, so the compiler cannot get rid of the residual code because it is not known ahead of time if it is needed.

I cannot comment on whether the inputs being a power of 2 is common or uncommon. However, when it is a power of 2, we can optimize the code in two ways:
- Replace direct memory accesses with subgroup loads and stores
- Avoid any calculations of the residual and potential diverging branches

These optimizations provide a performance improvement that I believe is significant enough to specialize this code.